base/download.jl

```diff
@@ -19,29 +19,55 @@ if Sys.iswindows()
             end
             pipeline_error(proc)
         end
-        filename
+        return filename
     end
-else
-    function download(url::AbstractString, filename::AbstractString)
-        if Sys.isapple() && Sys.isexecutable("/usr/bin/curl")
-            run(`/usr/bin/curl -g -L -f -o $filename $url`) # issue #30956
-        elseif Sys.which("curl") !== nothing
-            run(`curl -g -L -f -o $filename $url`)
-        elseif Sys.which("wget") !== nothing
-            try
-                run(`wget -O $filename $url`)
-            catch
-                rm(filename, force=true)  # wget always creates a file
-                rethrow()
-            end
-        elseif Sys.which("fetch") !== nothing
-            run(`fetch -f $filename $url`)
-        else
-            error("no download agent available; install curl, wget, or fetch")
+end
+
+function find_curl()
+    if Sys.isapple() && Sys.isexecutable("/usr/bin/curl")
+        "/usr/bin/curl"
+    elseif Sys.iswindows() && Sys.isexecutable(joinpath(get(ENV, "SYSTEMROOT", "C:\\Windows"), "System32\\curl.exe"))
+        joinpath(get(ENV, "SYSTEMROOT", "C:\\Windows"), "System32\\curl.exe")
+    elseif Sys.which("curl") !== nothing
+        "curl"
+    else
+        nothing
+    end
+end
+
+function download_curl(curl_exe::AbstractString, url::AbstractString, filename::AbstractString)
+    out, err = Pipe(), Pipe()
+    process = run(pipeline(`$curl_exe -s -S -g -L -o $filename $url`, stdout=out, stderr=err), wait=false)
+    close(out.in); close(err.in)
+    wait(process)
+    stderr = read(err, String)
```

Comment:
This ordering of events is a setup for getting deadlocks. Easiest solution is to make it the intended type up front: `err = PipeBuffer()`

Also, what happened to `out`? Did you mean to pass `stdout=stdout` or `stdout=devnull` perhaps?

Reply:
I don't need stdout. It should be  `stdout=devnull`

Can you check the latest commit. Not sure how to use `PipeBuffer()` for the err since it's not readable?

Reply:
```julia
julia> b = PipeBuffer(); success(pipeline(`cat VERSION`, stdout=b))
Process(`cat VERSION`, ProcessExited(0))

julia> read(b, String)
"1.2.0-DEV\n"
```

Reply:
Thanks fixed.